### PR TITLE
Remove type alias definition on inline

### LINKED ISSUE #13079: `inline_type_alias_uses` should remove alias definition
Follow-up from #13036 / #10881.

### PR DESCRIPTION
**Pull request by ice1000**
Fix #13079

### CODE REVIEW

**Comment by ice1000** [comment]
I'm having trouble implementing the removal of `use` statements. With the latest code I get very weird indels generated by code that I don't know where they are -- I failed to setup lldb (breakpoints never hit), and backtrace library is unstable so I can't use it either.

Rust debugging is too difficult for me. I'm requesting help!

File: crates/ide-assists/src/handlers/inline_type_alias.rs [hunk 1]
```diff
@@ -84,6 +84,9 @@ pub(crate) fn inline_type_alias_uses(acc: &mut Assists, ctx: &AssistContext<'_>)
             for (file_id, refs) in usages.into_iter() {
                 inline_refs_for_file(file_id, refs);
             }
+
+            builder.edit_file(ctx.file_id());
+            builder.delete(ast_alias.syntax().text_range());
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/inline_type_alias.rs:89]: You'll have to extract the edits for `ctx.file_id()` out of `usages` and do the removal/replacement in one go like is done in `inline_call`. reason being if you `edit_file` for file `0`, then for `1` and then for `0` again you discard the previous edits.

File: crates/ide-assists/src/handlers/inline_type_alias.rs [hunk 2]
```diff
@@ -62,16 +66,12 @@ pub(crate) fn inline_type_alias_uses(acc: &mut Assists, ctx: &AssistContext<'_>)
...
-                        }
-                        _ => None,
-                    })
-                    .collect();
+                let (path_types, path_type_uses) =
+                    split_refs_and_uses(builder, refs, |path_type| {
+                        path_type.syntax().ancestors().nth(3).and_then(ast::PathType::cast)
+                    });
 
+                path_type_uses.iter().for_each(remove_path_if_in_use_stmt);
```
Comment (Veykril) [inline crates/ide-assists/src/handlers/inline_type_alias.rs:74]: This fails, we are looking explicilty for `ast::PathType` usages, but in imports there are no `ast::PathType`s, only `ast::Path`s

Comment (ice1000) [reply crates/ide-assists/src/handlers/inline_type_alias.rs:74]: image

I think the types are correct ;-) did the variable name confuse you?

File: crates/ide-assists/src/handlers/inline_type_alias.rs (hunk 1 above)
Comment (ice1000) [reply crates/ide-assists/src/handlers/inline_type_alias.rs]: Could you please check my latest changes?

**Comment by ice1000** [comment]
Still fails. The stack traces look like

```
   3: text_edit::assert_disjoint_or_equal
   4: text_edit::TextEditBuilder::indel
   5: text_edit::TextEditBuilder::delete
   6: syntax::algo::TreeDiff::into_text_edit
   7: ide_db::source_change::SourceChangeBuilder::commit
   8: ide_db::source_change::SourceChangeBuilder::finish
   9: ide_assists::assist_context::Assists::add_impl
  10: ide_assists::assist_context::Assists::add
  11: ide_assists::handlers::inline_type_alias::inline_type_alias_uses
```

I'm unsure why is that happening

File: crates/ide-assists/src/handlers/inline_type_alias.rs (hunk 2 above)
Comment (Veykril) [reply crates/ide-assists/src/handlers/inline_type_alias.rs:75]: Ah apoligies, I misunderstood how `split_refs_and_uses` worked, yes this is correct then

**Comment by Veykril** [comment]
I believe the problem you are running into is that you are `split_refs_and_uses` + `remove_path_if_in_use_stmt` does removal via mutable syntax nodes, this creates a diff of the tree that then gets applied as text edits, this diff in this case is very disruptive and overlaps with your manual `replace` calls causing the panic. In general mixing these two APIs (manual `replace`, `delete` etc calls on `builder` with text ranges and using mutable syntax nodes) does not work out well.

**Comment by ice1000** [comment]
@Veykril is it possible to use `ted::replace` to replace a tree node with an unparsed text? (which is the return value of `to_text` here)

**Comment by ice1000** [comment]
@Veykril I'm done! :)

**Comment by ice1000** [comment]
ping @Veykril

**Pull request merged**